nutils/function.py

```diff
@@ -18,2906 +18,2514 @@
-    try:
-      index = arguments.index(bound.arguments) # by using index, arguments need not be hashable
-    except ValueError:
-      index = -1
-      cache = weakref.WeakKeyDictionary()
+    if isinstance(__value, Array):
+      value = __value
+    elif numeric.isnumber(__value) or numeric.isarray(__value):
+      value = _Constant(__value)
+    elif isinstance(__value, (list, tuple)):
+      value = stack(__value, axis=0)
     else:
-      cache = caches[index]
-    if index != 0: # function arguments are not the most recent (possibly new)
-      if index > 0 or len(arguments) >= lru:
-        caches.pop(index) # pop matching (or oldest) item
-        arguments.pop(index)
-      caches.insert(0, cache) # insert popped (or new) item to front
-      arguments.insert(0, bound.arguments)
-
-    fstack = [target] # stack of unprocessed objects and command tokens
-    rstack = [] # stack of processed objects
-    _stack = fstack if recursive else rstack
-
-    while fstack:
-      obj = fstack.pop()
-
-      if obj is recreate:
-        args = [rstack.pop() for obj in range(fstack.pop())]
-        f = fstack.pop()
-        r = f(*args)
-        if depthfirst:
-          newr = func(r, *funcargs, **funckwargs)
-          if newr is not None:
-            _stack.append(newr)
-            continue
-        rstack.append(r)
-        continue
-
-      if obj is remember:
-        cache[fstack.pop()] = rstack[-1]
-        continue
-
-      if isinstance(obj, (tuple, list, dict, set, frozenset)):
-        if not obj:
-          rstack.append(obj) # shortcut to avoid recreation of empty container
-        else:
-          fstack.append(lambda *x, T=type(obj): T(x))
-          fstack.append(len(obj))
-          fstack.append(recreate)
-          fstack.extend(obj if not isinstance(obj, dict) else obj.items())
-        continue
-
-      try:
-        r = cache[obj]
-      except KeyError: # object can be weakly cached, but isn't
-        cache[obj] = pending
-        fstack.append(obj)
-        fstack.append(remember)
-      except TypeError: # object cannot be referenced or is not hashable
-        pass
-      else: # object is in cache
-        if r is pending:
-          pending_objs = [k for k, v in cache.items() if v is pending]
-          index = pending_objs.index(obj)
-          raise Exception('{}@replace caught in a circular dependence\n'.format(func.__name__) + Tuple(pending_objs[index:]).asciitree().split('\n', 1)[1])
-        rstack.append(r)
-        continue
-
-      if not depthfirst:
-        newr = func(obj, *funcargs, **funckwargs)
-        if newr is not None:
-          _stack.append(newr)
-          continue
-
-      try:
-        f, args = obj.__reduce__()
-      except: # obj cannot be reduced into a constructor and its arguments
-        rstack.append(obj)
-      else:
-        fstack.append(f)
-        fstack.append(len(args))
-        fstack.append(recreate)
-        fstack.extend(args)
+      raise ValueError
```

Comment:
Let's add a message here such as ValueError('cannot convert {} to Array'.format(type(__value).__name). Also perhaps we can switch to single underscores for __value so it becomes printable in pdb.

Reply:
Done. Mypy treats parameters with double underscores as positional only (see https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#functions , function `quux`).

Reply:
Ah yes, that was it.